trinity/protocol/common/peer.py

```diff
@@ -0,0 +1,253 @@
+                break
+            elif start_block > vm_class.dao_fork_block_number:
+                # VM comes after the fork, so stop checking
+                break
+
+            start_block = vm_class.dao_fork_block_number - 1
+
+            try:
+                headers = await self.requests.get_block_headers(  # type: ignore
+                    start_block,
+                    max_headers=2,
+                    reverse=False,
+                    timeout=CHAIN_SPLIT_CHECK_TIMEOUT,
+                )
+
+            except (TimeoutError, PeerConnectionLost) as err:
+                raise DAOForkCheckFailure(
+                    f"Timed out waiting for DAO fork header from {self}: {err}"
+                ) from err
+            except ValidationError as err:
+                raise DAOForkCheckFailure(
+                    f"Invalid header response during DAO fork check: {err}"
+                ) from err
+
+            if len(headers) != 2:
+                raise DAOForkCheckFailure(
+                    f"{self} failed to return DAO fork check headers"
+                )
+            else:
+                parent, header = headers
+
+            try:
+                vm_class.validate_header(header, parent, check_seal=True)
+            except ValidationError as err:
+                raise DAOForkCheckFailure(f"{self} failed DAO fork check validation: {err}")
+
+    @property
+    async def genesis(self) -> BlockHeader:
+        genesis_hash = await self.wait(
+            self.headerdb.coro_get_canonical_block_hash(BlockNumber(GENESIS_BLOCK_NUMBER)))
+        return await self.wait(self.headerdb.coro_get_block_header_by_hash(genesis_hash))
+
+    @property
+    async def _local_chain_info(self) -> ChainInfo:
+        genesis = await self.genesis
+        head = await self.wait(self.headerdb.coro_get_canonical_head())
+        total_difficulty = await self.headerdb.coro_get_score(head.hash)
+        return ChainInfo(
+            block_number=head.block_number,
+            block_hash=head.hash,
+            total_difficulty=total_difficulty,
+            genesis_hash=genesis.hash,
+        )
+
+
+class EVMPeerPool(BasePeerPool):
+    peer_class: Type[EVMPeer]
+    # ignore type here to force the class to use EVMPeer as the base peer
+    # class.
+    connected_nodes: Dict[Node, EVMPeer]  # type: ignore
+
+    def __init__(self,
+                 headerdb: 'BaseAsyncHeaderDB',
+                 network_id: int,
+                 vm_configuration: Tuple[Tuple[int, Type[BaseVM]], ...],
+                 privkey: eth_keys_datatypes.PrivateKey,
+                 max_peers: int = DEFAULT_MAX_PEERS,
+                 token: CancelToken = None,
+                 ) -> None:
+        super().__init__(privkey=privkey, max_peers=max_peers, token=token)
+        self.headerdb = headerdb
+        self.network_id = network_id
+        self.vm_configuration = vm_configuration
+
+    def create_peer(self,
```

Comment:
I'm going to try updating this to remove the `create_peer` in favor of a composition based approach, adding a `PeerFactory` which is responsible for instantiating peers.  I think it'll make the types cleaner and I'm hoping it makes some of the code here cleaner.